org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend

```diff
@@ -3749,6 +4010,14 @@ class ServerCapabilities {
 	 */
 	Either<Boolean, StaticRegistrationOptions> selectionRangeProvider
 
+	/**
+	 * The server provides semantic tokens support.
+	 *
+	 * Since 3.16.0
+	 */
+	@Beta
+	Either<SemanticTokensOptions, SemanticTokensRegistrationOptions> semanticTokensProvider
```

Comment:
This won't work because both alternatives are objects, so the EitherTypeAdapter is not able to decide whether to parse the left or right alternative. You can solve this with an additional type adapter that configures an EitherTypeAdapter with two PropertyChecker instances. See the `org.eclipse.lsp4j.adapters` package for examples.

Please add a parser test to verify that parsing currently fails, and that the new type adapter fixes the test.

Reply:
Alternatively, you could check whether it would be viable to merge both alternatives into a single class.

Reply:
Thanks, I'll try to fix that

Reply:
Merged into `SemanticTokensWithRegistrationOptions`